--- samd51_vanilla/hal/utils/include/utils_assert.h	2017-09-20 22:33:52.000000000 -0700
+++ samd51/hal/utils/include/utils_assert.h	2017-09-19 13:07:30.000000000 -0700
@@ -72,7 +72,7 @@
 	if (!(condition))                                                                                                  \
 		__asm("BKPT #0");
 #else
-#define ASSERT_IMPL(condition, file, line) assert((condition), file, line)
+#define ASSERT_IMPL(condition, file, line) asf_assert((condition), file, line)
 #endif

 #else /* DEBUG */
@@ -95,7 +95,7 @@
  * \param[in] file File name
  * \param[in] line Line number
  */
-void assert(const bool condition, const char *const file, const int line);
+void asf_assert(const bool condition, const char *const file, const int line);

 #ifdef __cplusplus
 }
