--- samd51_vanilla/include/samd51g18a.h	2017-09-20 22:33:52.000000000 -0700
+++ samd51/include/samd51g18a.h 2017-09-19 13:07:30.000000000 -0700
@@ -511,7 +512,7 @@
  * \brief Configuration of the Cortex-M4 Processor and Core Peripherals
  */

-#define LITTLE_ENDIAN          1
+//#define LITTLE_ENDIAN          1
 #define __CM4_REV              1         /*!< Core revision r0p1 */
 #define __DEBUG_LVL            3         /*!< Full debug plus DWT data matching */
 #define __FPU_PRESENT          1         /*!< FPU present or not */
--- samd51_vanilla/include/samd51g19a.h 2017-09-20 22:33:52.000000000 -0700
+++ samd51/include/samd51g19a.h	2017-09-19 13:07:30.000000000 -0700
@@ -511,7 +512,7 @@
  * \brief Configuration of the Cortex-M4 Processor and Core Peripherals
  */

-#define LITTLE_ENDIAN          1
+//#define LITTLE_ENDIAN          1
 #define __CM4_REV              1         /*!< Core revision r0p1 */
 #define __DEBUG_LVL            3         /*!< Full debug plus DWT data matching */
 #define __FPU_PRESENT          1         /*!< FPU present or not */
--- samd51_vanilla/include/samd51j18a.h 2017-09-20 22:33:52.000000000 -0700
+++ samd51/include/samd51j18a.h	2017-09-19 13:07:30.000000000 -0700
@@ -529,7 +530,7 @@
  * \brief Configuration of the Cortex-M4 Processor and Core Peripherals
  */

-#define LITTLE_ENDIAN          1
+//#define LITTLE_ENDIAN          1
 #define __CM4_REV              1         /*!< Core revision r0p1 */
 #define __DEBUG_LVL            3         /*!< Full debug plus DWT data matching */
 #define __FPU_PRESENT          1         /*!< FPU present or not */
--- samd51_vanilla/include/samd51j19a.h 2017-09-20 22:33:52.000000000 -0700
+++ samd51/include/samd51j19a.h	2017-09-19 13:07:30.000000000 -0700
@@ -529,7 +530,7 @@
  * \brief Configuration of the Cortex-M4 Processor and Core Peripherals
  */

-#define LITTLE_ENDIAN          1
+//#define LITTLE_ENDIAN          1
 #define __CM4_REV              1         /*!< Core revision r0p1 */
 #define __DEBUG_LVL            3         /*!< Full debug plus DWT data matching */
 #define __FPU_PRESENT          1         /*!< FPU present or not */
--- samd51_vanilla/include/samd51j20a.h 2017-09-20 22:33:52.000000000 -0700
+++ samd51/include/samd51j20a.h	2017-09-19 13:07:30.000000000 -0700
@@ -529,7 +530,7 @@
  * \brief Configuration of the Cortex-M4 Processor and Core Peripherals
  */

-#define LITTLE_ENDIAN          1
+//#define LITTLE_ENDIAN          1
 #define __CM4_REV              1         /*!< Core revision r0p1 */
 #define __DEBUG_LVL            3         /*!< Full debug plus DWT data matching */
 #define __FPU_PRESENT          1         /*!< FPU present or not */
--- samd51_vanilla/include/samd51n19a.h 2017-09-20 22:33:52.000000000 -0700
+++ samd51/include/samd51n19a.h	2017-09-19 13:07:30.000000000 -0700
@@ -551,7 +552,7 @@
  * \brief Configuration of the Cortex-M4 Processor and Core Peripherals
  */

-#define LITTLE_ENDIAN          1
+//#define LITTLE_ENDIAN          1
 #define __CM4_REV              1         /*!< Core revision r0p1 */
 #define __DEBUG_LVL            3         /*!< Full debug plus DWT data matching */
 #define __FPU_PRESENT          1         /*!< FPU present or not */
--- samd51_vanilla/include/samd51n20a.h 2017-09-20 22:33:52.000000000 -0700
+++ samd51/include/samd51n20a.h	2017-09-19 13:07:30.000000000 -0700
@@ -551,7 +552,7 @@
  * \brief Configuration of the Cortex-M4 Processor and Core Peripherals
  */

-#define LITTLE_ENDIAN          1
+//#define LITTLE_ENDIAN          1
 #define __CM4_REV              1         /*!< Core revision r0p1 */
 #define __DEBUG_LVL            3         /*!< Full debug plus DWT data matching */
 #define __FPU_PRESENT          1         /*!< FPU present or not */
--- samd51_vanilla/include/samd51p19a.h 2017-09-20 22:33:52.000000000 -0700
+++ samd51/include/samd51p19a.h	2017-09-19 13:07:30.000000000 -0700
@@ -551,7 +552,7 @@
  * \brief Configuration of the Cortex-M4 Processor and Core Peripherals
  */

-#define LITTLE_ENDIAN          1
+//#define LITTLE_ENDIAN          1
 #define __CM4_REV              1         /*!< Core revision r0p1 */
 #define __DEBUG_LVL            3         /*!< Full debug plus DWT data matching */
 #define __FPU_PRESENT          1         /*!< FPU present or not */
--- samd51_vanilla/include/samd51p20a.h 2017-09-20 22:33:52.000000000 -0700
+++ samd51/include/samd51p20a.h	2017-09-19 13:07:30.000000000 -0700
@@ -551,7 +552,7 @@
  * \brief Configuration of the Cortex-M4 Processor and Core Peripherals
  */

-#define LITTLE_ENDIAN          1
+//#define LITTLE_ENDIAN          1
 #define __CM4_REV              1         /*!< Core revision r0p1 */
 #define __DEBUG_LVL            3         /*!< Full debug plus DWT data matching */
 #define __FPU_PRESENT          1         /*!< FPU present or not */
