--- samd51_vanilla/gcc/gcc/startup_samd51.c	2017-09-20 22:33:52.000000000 -0700
+++ samd51/gcc/gcc/startup_samd51.c	2017-09-19 13:07:30.000000000 -0700
@@ -314,7 +315,7 @@
 #endif

 /* Exception Table */
-__attribute__((section(".vectors"))) const DeviceVectors exception_table
+__attribute__((section(".vectors"), used)) const DeviceVectors exception_table
     = {

         /* Configure Initial Stack Pointer, using linker-generated symbols */
@@ -660,7 +661,7 @@
 #endif

 	/* Initialize the C library */
-	__libc_init_array();
+	//__libc_init_array();

 	/* Branch to main function */
 	main();
