--- samd21_vanilla/CMSIS/Include/core_cmInstr.h	2017-09-28 12:05:22.000000000 -0700
+++ samd21/CMSIS/Include/core_cmInstr.h	2017-09-19 13:07:29.000000000 -0700
@@ -171,7 +171,7 @@
 #define __BKPT(value)                       __breakpoint(value)
 
 
-#if       (__CORTEX_M >= 0x03) || (__CORTEX_SC >= 300)
+#if       (__CORTEX_M >= 0x03) || (defined(__CORTEX_SC) && __CORTEX_SC >= 300)
 
 /** \brief  Reverse bit order of value
 
@@ -538,7 +538,7 @@
 #define __BKPT(value)                       __ASM volatile ("bkpt "#value)
 
 
-#if       (__CORTEX_M >= 0x03) || (__CORTEX_SC >= 300)
+#if       (__CORTEX_M >= 0x03) || (defined(__CORTEX_SC) && __CORTEX_SC >= 300)
 
 /** \brief  Reverse bit order of value
 
