--- samd21_vanilla/hpl/gclk/hpl_gclk.c	2017-09-28 12:05:22.000000000 -0700
+++ samd21/hpl/gclk/hpl_gclk.c	2017-09-19 13:07:29.000000000 -0700
@@ -147,7 +147,7 @@
 	                               | CONF_GCLK_GEN_7_SRC
 	                               | GCLK_GENCTRL_ID(7));
 #endif
-#if CONF_GCLK_GEN_8_GENEN == 1
+#if defined(CONF_GCLK_GEN_8_GENEN) && CONF_GCLK_GEN_8_GENEN == 1
 	hri_gclk_write_GENDIV_reg(GCLK, GCLK_GENDIV_DIV(CONF_GCLK_GEN_8_DIV) | GCLK_GENDIV_ID(8));
 	hri_gclk_write_GENCTRL_reg(GCLK,
 	                           (CONF_GCLK_GEN_8_RUNSTDBY << GCLK_GENCTRL_RUNSTDBY_Pos)
