--- samd21_vanilla/include/samd21e18a.h  2017-09-28 12:05:22.000000000 -0700
+++ samd21/include/samd21e18a.h  2017-09-19 13:07:30.000000000 -0700
@@ -210,7 +210,7 @@
  * \brief Configuration of the Cortex-M0+ Processor and Core Peripherals
  */
 
-#define LITTLE_ENDIAN          1        
+// #define LITTLE_ENDIAN          1        
 #define __CM0PLUS_REV          1         /*!< Core revision r0p1 */
 #define __MPU_PRESENT          0         /*!< MPU present or not */
 #define __NVIC_PRIO_BITS       2         /*!< Number of bits used for Priority Levels */
--- samd21_vanilla/include/samd21g18a.h  2017-09-28 12:05:22.000000000 -0700
+++ samd21/include/samd21g18a.h  2017-09-19 13:07:30.000000000 -0700
@@ -214,7 +214,7 @@
  * \brief Configuration of the Cortex-M0+ Processor and Core Peripherals
  */
 
-#define LITTLE_ENDIAN          1        
+//#define LITTLE_ENDIAN          1        
 #define __CM0PLUS_REV          1         /*!< Core revision r0p1 */
 #define __MPU_PRESENT          0         /*!< MPU present or not */
 #define __NVIC_PRIO_BITS       2         /*!< Number of bits used for Priority Levels */
