--- samd21_vanilla/hpl/usb/hpl_usb.c	2017-09-28 12:05:22.000000000 -0700
+++ samd21/hpl/usb/hpl_usb.c	2017-09-19 13:07:29.000000000 -0700
@@ -300,7 +300,7 @@
 /** Endpoint cache buffer for OUT transactions (none-control) or SETUP/IN/OUT
  *  transactions (control). */
 #if CONF_USB_EP1_CACHE && CONF_USB_D_MAX_EP_N >= 1
-static uint32_t _usb_ep1_cache[_usb_align_up(CONF_USB_EP1_CACHE) / 4];
+uint32_t _usb_ep1_cache[_usb_align_up(CONF_USB_EP1_CACHE) / 4];
 #else
 #define _usb_ep1_cache NULL
 #endif
