--- samd21_vanilla/gcc/gcc/startup_samd21.c	2017-09-20 22:33:52.000000000 -0700
+++ samd21/gcc/gcc/startup_samd21.c	2017-09-19 13:07:30.000000000 -0700
@@ -104,7 +105,7 @@
 void I2S_Handler(void) __attribute__((weak, alias("Dummy_Handler")));
 
 /* Exception Table */
-__attribute__((section(".vectors"))) const DeviceVectors exception_table = {
+__attribute__((section(".vectors"), used)) const DeviceVectors exception_table = {

     /* Configure Initial Stack Pointer, using linker-generated symbols */
     .pvStack = (void *)(&_estack),
@@ -234,7 +234,7 @@
	NVMCTRL->CTRLB.bit.MANW = 1;

 	/* Initialize the C library */
-	__libc_init_array();
+	//__libc_init_array();

 	/* Branch to main function */
 	main();
