--- samd21_vanilla/hpl/sysctrl/hpl_sysctrl.c	2017-09-28 12:05:22.000000000 -0700
+++ samd21/hpl/sysctrl/hpl_sysctrl.c	2017-09-19 13:07:29.000000000 -0700
@@ -114,7 +114,7 @@
 
 #if CONF_OSCULP32K_CONFIG == 1
 	hri_sysctrl_write_OSCULP32K_reg(hw,
-#if OSC32K_OVERWRITE_CALIBRATION == 1
+#if CONF_OSC32K_OVERWRITE_CALIBRATION == 1
 	                                SYSCTRL_OSCULP32K_CALIB(CONF_OSCULP32K_CALIB) |
 #else
 	                                SYSCTRL_OSCULP32K_CALIB(calib) |
