--- samd21_vanilla/usb/device/usbdc.c	2017-09-28 12:05:22.000000000 -0700
+++ samd21/usb/device/usbdc.c	2017-09-19 13:07:30.000000000 -0700
@@ -739,7 +737,7 @@
  */
 int32_t usbdc_check_desces(struct usbdc_descriptors *desces)
 {
-#if CONF_USBD_HS_SP
+#ifdef CONF_USBD_HS_SP
 	int32_t rc;
 	if (desces->hs == NULL && desces->ls_fs == NULL) {
 		return ERR_NOT_FOUND;
--- samd21_vanilla/usb/device/usbdc.h	2017-09-28 12:05:22.000000000 -0700
+++ samd21/usb/device/usbdc.h	2017-09-19 13:07:30.000000000 -0700
@@ -105,7 +105,7 @@
 /** Describes the USB device core descriptors. */
 struct usbdc_descriptors {
 	struct usbd_descriptors *ls_fs;
-#if CONF_USBD_HS_SP
+#ifdef CONF_USBD_HS_SP
 	struct usbd_descriptors *hs;
 #endif
 };
